### DISCUSSION TITLE
assert a function raises

### DISCUSSION THREAD

**Discussion by jdunck** [microsoft/pyright#5408] [Ideas]
There seems to be special support for `raise` that allows following code to narrow the type based on the fact that we didn't raise.

I would like to see support for asserting that a function call raises, therefore offering the same narrowing as an explicit raise.

```python
def complicated_thing() -> int:
    return 1

def raises():
    raise ValueError()


def guarded_by_raise() -> int:
    try:
        x = complicated_thing()
    except Exception:
        raise
    return x + 1  # x: int


def guarded_by_exception_handler() -> int:
    try:
        x = complicated_thing()
    except Exception:
        raises()
    return x + 1  # currently - x: (int | Unbound); desired - x: int
```

I would like to be able to annotate `raises` here as a function that can be used for narrowing like the `raise` statement.

**Comment by hmc-cs-mdrissi** [comment]
Use NoReturn as return type will do it. Or Never. Both mean same thing and can be found in typing-extensions (or typing with new enough python version).

**Reply by erictraut** [reply]
As Mehdi said, the way to do this is to use a `NoReturn` return type annotation.

```python
from typing import NoReturn


def complicated_thing() -> int:
    return 1


def raises() -> NoReturn:
    raise ValueError()


def guarded_by_exception_handler() -> int:
    try:
        x = complicated_thing()
    except Exception:
        raises()
    return x + 1
```

**Comment by jdunck** [comment]
Thanks both. :)
